larpandoracontent/LArTwoDReco/TrackInEMShowerAlgorithm.cc

```diff
@@ -452,19 +413,59 @@ bool TrackInEMShowerAlgorithm::IsInLineSegment(const CartesianVector &lowerBound
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
-void TrackInEMShowerAlgorithm::RefineTracks(ClusterAssociation &clusterAssociation, TwoDSlidingFitResultMap &microSlidingFitResultMap,
-    TwoDSlidingFitResultMap &macroSlidingFitResultMap, ClusterVector &clusterVector, CaloHitVector &extrapolatedCaloHitVector) const
+void TrackInEMShowerAlgorithm::CreateMainTrack(const ClusterAssociation &clusterAssociation, const CaloHitVector &extrapolatedCaloHitVector, const ClusterToCaloHitListMap &clusterToCaloHitListMap, TwoDSlidingFitResultMap &microSlidingFitResultMap, TwoDSlidingFitResultMap &macroSlidingFitResultMap, ClusterVector &clusterVector) const
 {
-    clusterAssociation.SetUpstreamCluster(this->RefineTrack(clusterAssociation.GetUpstreamCluster(), clusterAssociation.GetUpstreamMergePoint(), extrapolatedCaloHitVector, true,
-        microSlidingFitResultMap, macroSlidingFitResultMap, clusterVector));
-    clusterAssociation.SetDownstreamCluster(this->RefineTrack(clusterAssociation.GetDownstreamCluster(), clusterAssociation.GetDownstreamMergePoint(), extrapolatedCaloHitVector, false,
-        microSlidingFitResultMap, macroSlidingFitResultMap, clusterVector));
+    ClusterList remnantClusterList;
+
+    // Remove off axis hits from main track cluster pair
+    const Cluster *const pMainTrackCluster(this->RemoveOffAxisHitsFromTrack(clusterAssociation.GetUpstreamCluster(), clusterAssociation.GetUpstreamMergePoint(),
+        true, extrapolatedCaloHitVector, remnantClusterList, microSlidingFitResultMap, macroSlidingFitResultMap, clusterVector));
+    const Cluster *const pClusterToDelete(this->RemoveOffAxisHitsFromTrack(clusterAssociation.GetDownstreamCluster(), clusterAssociation.GetDownstreamMergePoint(),
+        false, extrapolatedCaloHitVector, remnantClusterList, microSlidingFitResultMap, macroSlidingFitResultMap, clusterVector));
+
+    // Merge main track cluster pair together
+    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::MergeAndDeleteClusters(*this, pMainTrackCluster, pClusterToDelete));
+
+    // Determine the shower clusters that contain hits which belong to the main track
+    ClusterVector showerClustersToFragment;
+    for (const auto &entry : clusterToCaloHitListMap)
+        showerClustersToFragment.push_back(entry.first);
+
+    std::sort(showerClustersToFragment.begin(), showerClustersToFragment.end(), LArClusterHelper::SortByNHits);
+
+    // Remove hits from inbetween shower
+    if (!showerClustersToFragment.empty())
```

Comment:
This check not needed, and can just keep the loop below (where nothing will happen if container is empty).